lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+  }
+}
+
+void PSVContentVerifier::VerifySignatures(unsigned ValMajor,
+                                          unsigned ValMinor) {
+  bool i1ToUnknownCompat = DXIL::CompareVersions(ValMajor, ValMinor, 1, 5) < 0;
+  // Verify input signature
+  VerifySignature(DM.GetInputSignature(), PSV.GetInputElement0(0),
+                  PSV.GetSigInputElements(), "SigInput", i1ToUnknownCompat);
+  // Verify output signature
+  VerifySignature(DM.GetOutputSignature(), PSV.GetOutputElement0(0),
+                  PSV.GetSigOutputElements(), "SigOutput", i1ToUnknownCompat);
+  // Verify patch constant signature
+  VerifySignature(DM.GetPatchConstOrPrimSignature(),
+                  PSV.GetPatchConstOrPrimElement0(0),
+                  PSV.GetSigPatchConstOrPrimElements(),
+                  "SigPatchConstantOrPrim", i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignature(const DxilSignature &Sig,
+                                         PSVSignatureElement0 *Base,
+                                         unsigned Count, std::string Name,
+                                         bool i1ToUnknownCompat) {
+  if (Count != Sig.GetElements().size()) {
+    EmitMismatchError(Name + "Elements", std::to_string(Count),
+                      std::to_string(Sig.GetElements().size()));
+    return;
+  }
+
+  // Verify each element in DxilSignature.
+  const PSVStringTable &StrTab = PSV.GetStringTable();
+  const PSVSemanticIndexTable &IndexTab = PSV.GetSemanticIndexTable();
+  for (unsigned i = 0; i < Count; i++)
+    VerifySignatureElement(Sig.GetElement(i),
+                           PSV.GetRecord<PSVSignatureElement0>(
+                               Base, PSV.GetSignatureElementSize(), Count, i),
+                           StrTab, IndexTab, Name, i1ToUnknownCompat);
+}
+
+void PSVContentVerifier::VerifySignatureElement(
+    const DxilSignatureElement &SE, PSVSignatureElement0 *PSVSE0,
+    const PSVStringTable &StrTab, const PSVSemanticIndexTable &IndexTab,
+    std::string Name, bool i1ToUnknownCompat) {
+  // Find the signature element in the set.
+  PSVSignatureElement0 ModulePSVSE0;
+  InitPSVSignatureElement(ModulePSVSE0, SE, i1ToUnknownCompat);
+
+  // Check the Name and SemanticIndex.
+  bool Mismatch = false;
+  const std::vector<uint32_t> &SemanticIndexVec = SE.GetSemanticIndexVec();
+  llvm::ArrayRef<uint32_t> PSVSemanticIndexVec(
+      IndexTab.Get(PSVSE0->SemanticIndexes), PSVSE0->Rows);
+  if (SemanticIndexVec.size() == PSVSemanticIndexVec.size())
+    Mismatch |= memcmp(PSVSemanticIndexVec.data(), SemanticIndexVec.data(),
+                       SemanticIndexVec.size() * sizeof(uint32_t)) != 0;
+  else
+    Mismatch = true;
+
+  ModulePSVSE0.SemanticIndexes = PSVSE0->SemanticIndexes;
+
+  PSVSignatureElement PSVSE(StrTab, IndexTab, PSVSE0);
+  if (SE.IsArbitrary())
+    Mismatch |= strcmp(PSVSE.GetSemanticName(), SE.GetName());
+  else
+    Mismatch |= PSVSE0->SemanticKind != static_cast<uint8_t>(SE.GetKind());
```

Comment:
Are both of these conditions tested? It looks like semantickind is not copied over.

Reply:
SemanticKind was initialized in InitPSVSignatureElement.
Only SemanticName and SemanticSemanticIndexes are copied because the data in PSVSignatureElement0 are just index.
PSVStringTableReorder will cover the semantic name path.
SemanticKind is not tested mismatch directly. But not mismatch was tested.